### PR TITLE
hypervisor: emulator: emulate CMP

### PR DESCRIPTION
**Pull request by liuw**
Unfortunately Rust stable does not yet have inline ASM support the flag
calculation will have to be implemented in software.

Signed-off-by: Wei Liu <liuwe@microsoft.com>

### CODE REVIEW

**Comment by liuw** [comment]
This series now contains a few more patches to reduce code repetition.

File: hypervisor/src/arch/x86/emulator/instructions/mod.rs [hunk 1]
```diff
@@ -47,6 +47,89 @@ macro_rules! imm_op {
 
 pub mod mov;
 
+fn get_op<T: CpuStateManager>(
+    insn: &Instruction,
+    n: u32,
```
Comment (sameo) [inline hypervisor/src/arch/x86/emulator/instructions/mod.rs:19]: Could we use more explicit naming, like e.g. `op_index`?

**Review by rbradford — APPROVED** [review]
Looks good. We can ignore the nightly failure. A PR is already open to fix it in rust-lang/rust

File: hypervisor/src/arch/x86/emulator/instructions/mod.rs [hunk 2]
```diff
@@ -47,6 +47,89 @@ macro_rules! imm_op {
...
+    state: &mut T,
+    platform: &mut dyn PlatformEmulator<CpuState = T>,
+    size: usize,
```
Comment (sameo) [inline hypervisor/src/arch/x86/emulator/instructions/mod.rs:22]: Ditto, can we use `op_size` and maybe move it next to `op_index` in the argument list?

File: hypervisor/src/arch/x86/emulator/instructions/mov.rs [hunk 3]
```diff
@@ -27,27 +26,18 @@ macro_rules! mov_rm_r {
...
-                }
-
-                k => return Err(EmulationError::InvalidOperand(anyhow!("{:?}", k))),
-            }
+            set_op(
+                &insn,
+                0,
+                state,
+                platform,
+                std::mem::size_of::<$bound>(),
```
Comment (sameo) [inline hypervisor/src/arch/x86/emulator/instructions/mov.rs:37]: This is already present in the code and should not be fixed with this PR, but we should have more consistent macros arguments names, and switch to `bound` for all of them.

File: hypervisor/src/arch/x86/emulator/instructions/cmp.rs [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+const SF_SHIFT: usize = 7;
+const OF_SHIFT: usize = 11;
+
+const CF: u64 = 1 << CF_SHIFT;
+const PF: u64 = 1 << PF_SHIFT;
+const AF: u64 = 1 << AF_SHIFT;
+const ZF: u64 = 1 << ZF_SHIFT;
+const SF: u64 = 1 << SF_SHIFT;
+const OF: u64 = 1 << OF_SHIFT;
+
```
Comment (sameo) [inline hypervisor/src/arch/x86/emulator/instructions/cmp.rs:33]: Maybe this is something that we should move to `hypervisor/src/arch/x86/mod.rs` or `hypervisor/src/arch/x86/regs.rs` as future emulated instructions may benefit from it.

File: hypervisor/src/arch/x86/emulator/instructions/cmp.rs [hunk 5]
```diff
@@ -241,22 +241,14 @@ mod tests {
...
     use crate::arch::x86::emulator::mock_vmm::*;
 
-    macro_rules! hashmap {
-        ($( $key: expr => $val: expr ),*) => {{
-            let mut map = ::std::collections::HashMap::new();
-            $( map.insert($key, $val); )*
-                map
-        }}
-    }
-
```
Comment (sameo) [inline hypervisor/src/arch/x86/emulator/instructions/cmp.rs]: For the record, `hasmap!` could be exported as a macro (`macro_export`) but it then is moved to the root of the crate. Having to write `use crate::hasmap` is not the prettiest and it's a little awkward to access it from the crate's root. So I'm ok with your change here.

**Review by sameo — APPROVED** [review]
I really only have nits or general comments. This looks really good, thanks for the work @liuw .

**Comment by sameo** [comment]
> Looks good. We can ignore the nightly failure. A PR is already open to fix it in rust-lang/rust

@rbradford Should we disable nightly builds until that's fixed in rust-lang/rust?

**Comment by sameo** [comment]
> > Looks good. We can ignore the nightly failure. A PR is already open to fix it in rust-lang/rust
> 
> @rbradford Should we disable nightly builds until that's fixed in rust-lang/rust?

nm, #2039 should do it.

File: hypervisor/src/arch/x86/emulator/instructions/cmp.rs (hunk 4 above)
Comment (liuw) [reply hypervisor/src/arch/x86/emulator/instructions/cmp.rs:33]: I can do this in a follow-up PR.

Comment (liuw) [reply hypervisor/src/arch/x86/emulator/instructions/cmp.rs]: Actually I've pushed a new patch to move the bits to `regs.rs`

**Pull request merged**